crates/control/src/kinematics_provider.rs

```diff
@@ -24,7 +36,85 @@ impl KinematicsProvider {
         Ok(Self {})
     }
 
-    pub fn cycle(&mut self, _context: CycleContext) -> Result<MainOutputs> {
-        Ok(MainOutputs::default())
+    pub fn cycle(&mut self, context: CycleContext) -> Result<MainOutputs> {
+        let joints = &context.sensor_data.positions;
+        // head
+        let neck_to_robot = neck_to_robot(&joints.head);
+        let head_to_robot = neck_to_robot * head_to_neck(&joints.head);
+        // torso
+        let torso_to_robot = Isometry3::from(RobotDimensions::TORSO_TO_ROBOT);
+        // left arm
+        let left_shoulder_to_robot = left_shoulder_to_robot(&joints.left_arm);
+        let left_upper_arm_to_robot =
+            left_shoulder_to_robot * left_upper_arm_to_left_shoulder(&joints.left_arm);
+        let left_elbow_to_robot =
+            left_upper_arm_to_robot * left_elbow_to_left_upper_arm(&joints.left_arm);
+        let left_forearm_to_robot =
+            left_elbow_to_robot * left_forearm_to_left_elbow(&joints.left_arm);
+        let left_wrist_to_robot =
+            left_forearm_to_robot * left_wrist_to_left_forearm(&joints.left_arm);
+        // right arm
+        let right_shoulder_to_robot = right_shoulder_to_robot(&joints.right_arm);
+        let right_upper_arm_to_robot =
+            right_shoulder_to_robot * right_upper_arm_to_right_shoulder(&joints.right_arm);
+        let right_elbow_to_robot =
+            right_upper_arm_to_robot * right_elbow_to_right_upper_arm(&joints.right_arm);
+        let right_forearm_to_robot =
+            right_elbow_to_robot * right_forearm_to_right_elbow(&joints.right_arm);
+        let right_wrist_to_robot =
+            right_forearm_to_robot * right_wrist_to_right_forearm(&joints.right_arm);
+        // left leg
```

Comment:
These comments may hint that your modularity is not well designed. Maybe group them into structs or functions to get rid of the comments and let your code document what belongs together?

Reply:
Uhh... You're right. But I vote to postpone such improvements to future work and open an issue. This again traces through to the layout of `RobotKinematics`.